### ISSUE TITLE
should the itemSource getter return the original Array or the converted ItemSource?

### ISSUE THREAD

**Issue by valdrinkoshi** [valdrinkoshi/virtual-scroller#93]
virtual-scroller allows to set an Array as item source, and it will convert it to an ItemSource instance via ItemSource.fromArray().
The itemSource getter returns the converted instance instead of the original array. This might cause difficulties in checking for identity.
Should we:
1) keep it as is: accept Array in the setter, return ItemSource in the getter 
1) accept Array in the setter and return same Array in the getter, while using the generated ItemSource internally
1) accept only ItemSource instances in the virtual-scroller setter

I think 2) would work best

**Comment by domenic** [comment]
So this kind of conversion is fairly common on the web platform. For example, anything that accepts a string, will convert an object to a string, store the string internally, and return the converted string from the getter. Having the getter return something different than what you set is not a big deal, in general.

There aren't as many examples for non-primitives, but we're considering something very similar for array / frozenarray: https://github.com/WICG/aom/issues/102#issuecomment-387927864. You can also make an argument that the various APIs that accept an array as an _argument_, then return back a copy of it, are kind of similar.

Finally, I think having .itemSource always return the same type of object, an ItemSource, is a good property. 2) has it sometimes return an array, and sometimes an ItemSource, which could be hard to program against.

So in short I'm OK-ish with the current behavior. This part is less clearly correct though:

```js
const arr = [];
scroller.itemSource = arr;

arr.push(5);
scroller.itemSource = arr; // does NOT call itemsChanged()

scroller.itemSource = [6]; // DOES call itemsChanged()
```

It's that behavior which might be worth changing.

/cc @ojanvafai since he was interested in this.

**Comment by valdrinkoshi** [comment]
Uhm, also this is not so clear, I don't know if that would be expected:
```js
const arr = [];

const source = ItemSource.fromArray(arr);
scroller.itemSource = source;

arr.push(1);
scroller.itemSource = source; // does NOT call itemsChanged()

scroller.itemSource = arr; // DOES call itemsChanged()

scroller.itemSource = source; // DOES call itemsChanged()
```

**Comment by ojanvafai** [comment]
Is it unreasonable to make setting itemSource unconditionally call itemsChanged?

**Comment by valdrinkoshi** [comment]
Discussed with @domenic, we think avoiding identity checks in the setter and always returning ItemSource instance in the getter is the way to go 👌
